### PR TITLE
Hotlinking: Perform CORS proxy requirement check

### LINKED ISSUE #9367: CORS Proxy: Perform preflight request
## Feature Description

Ideally in the editor we perform a (CORS preflight?) request first before trying to pass media requests through our proxy. So that we only proxy requests if needed due to performance considerations.

Something along these lines perhaps (untested):

```js
let shouldProxy = false;
try {
	  await fetch(
	      resourceUrl,
	      {
	          method: 'OPTIONS',
	          headers: {
	              'Accept': '*/*',
	              'Access-Control-Request-Method': 'GET',
	              'Access-Control-Request-Headers': {
	                  'Accept': '*/*',
	              }
	          },
	      }
	  )
} catch {
	shouldProxy = true
}
```

(untested)

## Acceptance Criteria

* There is a preflight request done when encountering an external resource the first time when loading the editor
* The response should be cached: ideally we only perform this check once per resource

### PR DESCRIPTION
**Pull request by swissspidy**
## Context

We only want to run hotlinked resources through our CORS proxy that actually need it.

## Summary

Performs a simple request upon hotlinking media insertion to see whether the 

## Relevant Technical Choices

There is a new permanent `needsProxy` flag on the `Resource` object that indicates the status.

## To-do

- Testing

## User-facing changes

N/A

## Testing Instructions

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1. Enable CORS proxy experiment
1. Insert images/videos from various sites for hotlinking, for example from Unsplash (supports CORS) as well as https://wp.stories.google/ (does not support it)
2. Verify that only images from https://wp.stories.google/ are proxied, the others aren't

### Does this PR have a security-related impact?

No

### Does this PR change what data or activity we track or use?

No

### Does this PR have a legal-related impact?

NO

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [ ] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [ ] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #9367

### CODE REVIEW

**Comment by miina** [comment]
Works as expected when testing. @swissspidy The PR is still marked as a Draft, is this just pending tests / testing and ready for review otherwise?

**Comment by swissspidy** [comment]
> is this just pending tests / testing and ready for review otherwise?

yes

File: packages/story-editor/src/components/library/panes/media/local/hotlink/useInsert.js [hunk 1]
```diff
@@ -79,8 +79,8 @@ function useInsert({ link, setLink, setErrorMsg, onClose }) {
       } = hotlinkData;
 
       try {
-        const proxiedUrl = getProxiedUrl({ isExternal: true }, link);
-        const resource = await getResourceFromUrl(proxiedUrl, type);
+        const proxiedUrl = getProxiedUrl({ needsProxy }, link);
```
Comment (miina) [inline packages/story-editor/src/components/library/panes/media/local/hotlink/useInsert.js]: Nit: Is there a reason to call `getProxiedUrl` at all here if `needsProxy` is `false`?

**Comment by spacedmonkey** [comment]
@swissspidy Have you look into why unit tests seem to failing here. I don't want to approve this until those are passing.

**Comment by swissspidy** [comment]
@spacedmonkey It's an issue on `main` that I am looking into. Not an issue with this PR. It can be disregarded.

File: packages/story-editor/src/components/library/panes/media/local/hotlink/useInsert.js (hunk 1 above)
Comment (swissspidy) [reply packages/story-editor/src/components/library/panes/media/local/hotlink/useInsert.js]: Good point. This can be simplified.

Comment (spacedmonkey) [inline packages/story-editor/src/components/library/panes/media/local/hotlink/useInsert.js]:
```suggestion
        const proxiedUrl = needsProxy ? getProxiedUrl({ needsProxy }, link) : link;
```
How about?

**Review by spacedmonkey — APPROVED** [review]
One nit pic, otherwise it looks good.

File: packages/story-editor/src/components/library/panes/media/local/karma/hotlink.karma.js [hunk 2]
```diff
@@ -79,6 +79,10 @@ describe('Embedding hotlinked media', () => {
   });
 
   it('should insert a new media element from valid url', async () => {
+    spyOn(window, 'fetch').and.callFake(() => {
+      return true;
+    });
```
Comment (swissspidy) [inline packages/story-editor/src/components/library/panes/media/local/karma/hotlink.karma.js:84]: Not sure this is really working 🤔

**Pull request merged**